mcstasscript/interface/plotter.py

```diff
@@ -0,0 +1,650 @@
+import numpy as np
+import matplotlib
+import matplotlib.pyplot as plt
+import matplotlib.animation as animation
+from matplotlib.colors import BoundaryNorm
+from matplotlib.ticker import MaxNLocator
+
+from mcstasscript.data.data import McStasMetaData
+from mcstasscript.data.data import McStasPlotOptions
+from mcstasscript.data.data import McStasData
+
+
+class make_plot:
+    """
+    make_plot plots contents of McStasData objects
+
+    Plotting is controlled through options assosciated with the
+    McStasData objects.
+
+    If a list is given, the plots appear individually.
+    """
+
+    def __init__(self, data_list, **kwargs):
+        """
+        plots McStasData, single object or list of McStasData
+
+        The options concerning plotting are stored with the data
+
+        Parameters
+        ----------
+        data_list : McStasData or list of McStasData
+            McStasData to be plotted
+        """
+
+        # Relevant options:
+        #  select colormap
+        #  show / hide colorbar
+        #  custom title / label
+        #  color of 1d plot
+        #  overlay several 1d
+        #  log scale (orders of magnitude)
+        #  compare several 1d
+        #  compare 2D
+
+        if isinstance(data_list, McStasData):
+            # Only a single element, put it in a list for easier syntax later
+            data_list = [data_list]
+
+        number_of_plots = len(data_list)
+
+        if "fontsize" in kwargs:
+            plt.rcParams.update({'font.size': kwargs["fontsize"]})
+
+        print("number of elements in data list = " + str(len(data_list)))
+
+        index = -1
+        for data in data_list:
+            index = index + 1
+
+            print("Plotting data with name " + data.metadata.component_name)
+            if type(data.metadata.dimension) == int:
+                fig = plt.figure(0)
+
+                x_axis_mult = data.plot_options.x_limit_multiplier
+                # print(data.T)
+                x = data.xaxis*x_axis_mult
+                y = data.Intensity
+                y_err = data.Error
+
+                #(fig, ax0) = plt.errorbar(x, y, yerr=y_err)
+                plt.errorbar(x, y, yerr=y_err)
+
+                ax0 = plt.gca()
+
+                if data.plot_options.log:
+                    ax0.set_yscale("log", nonposy='clip')
+
+                ax0.set_xlim(data.metadata.limits[0]*x_axis_mult,
+                             data.metadata.limits[1]*x_axis_mult)
+
+                # Add a title
+                plt.title(data.metadata.title)
+
+                # Add axis labels
+                plt.xlabel(data.metadata.xlabel)
+                plt.ylabel(data.metadata.ylabel)
+
+                if data.plot_options.custom_xlim_left:
+                    ax0.set_xlim(left=data.plot_options.left_lim)
+
+                if data.plot_options.custom_xlim_right:
+                    ax0.set_xlim(right=data.plot_options.right_lim)
+
+            elif len(data.metadata.dimension) == 2:
```

Comment:
I think I would try to split this into 2 different functions, one that plots 1D data, and another that makes 2d images.
Would be more readable and maintainable.